src/remarx/get_token_embeddings.py

```diff
@@ -0,0 +1,301 @@
+be changed to another pretrained model. Since term instances and sentences are
+unlikely to have a one-to-one correspondence, a metadata file (CSV) identifying
+each embedding is also generated.
+
+See https://huggingface.co/docs/transformers/main/en/model_doc/mt5 for more on MT5.
+
+Example:
+
+    python get_token_embeddings sents.jsonl Kapital output_dir output_stem
+
+"""
+
+import argparse
+import csv
+import pathlib
+import re
+import sys
+from typing import cast
+
+import numpy as np
+import numpy.typing as npt
+import orjsonl
+from tqdm import tqdm
+from transformers import (
+    FeatureExtractionPipeline,
+    PreTrainedTokenizer,
+    T5Tokenizer,
+    pipeline,
+)
+
+
+def normalize_text(text: str) -> str:
+    """
+    Normalizes text so that it will align with its tokenized version.
+    """
+    # Remove trailing whitespace
+    result_text = text.rstrip()
+    # Replace consecutive whitespace with a space
+    result_text = " ".join(result_text.split())
+    # Add leading space
+    result_text = " " + result_text
+    return result_text
+
+
+def get_term_spans(text: str, term: str) -> list[tuple[int, int]]:
+    """
+    Finds all instances of the provided term within the input text.
+    Returns a list of span start and end indices for each identified instance.
+    """
+    pattern = r"(\b|^)" + term + r"(\b|$)"
+    term_indices = []
+    for match in re.finditer(pattern, text):
+        term_indices.append(match.span())
+    return term_indices
+
+
+def get_subtoken_alignment(
+    tokenizer: PreTrainedTokenizer,
+    sentence: str,
+    spans: list[tuple[int, int]],
+) -> list[list[int]]:
+    """
+    For a sentence, determine the correspondence between each given span and
+    the given T5 tokenizer's subtokens. In cases where span and subtoken boundaries
+    are misaligned, all subtokens that contain chracters within the span are included.
+    Returns the span's subtoken positions as a list of lists of integers.
+    """
+    # Check that tokenized text aligns with original
+    subtokens = tokenizer.tokenize(sentence)
+    if len("".join(subtokens)) != len(sentence):
+        raise ValueError(
+            "Error: Tokenized sentence does not align with input. "
+            "Trying applying normalize_text first."
+        )
+
+    # Initialize span to subtoken correspondence
+    span2subtokens: list[list[int]] = [[] for x in spans]
+    # Calculate subtoken start indices
+    subtoken_starts: list[int] = []
+    for i, subtoken in enumerate(subtokens):
+        if i == 0:
+            start = 0
+        else:
+            start = subtoken_starts[i - 1] + len(subtoken)
+        subtoken_starts.append(start)
+
+    i_subtoken = 0
+    j_token = 0
+    while i_subtoken < len(subtokens) and j_token < len(spans):
+        sub_start = subtoken_starts[i_subtoken]
+        sub_end = (
+            subtoken_starts[i_subtoken + 1]
+            if i_subtoken < len(subtokens) - 1
```

Comment:
This looping logic is a bit hard to follow - I think it can be done in a single loop, esp. if you can assume spans are sorted (or sort them to guarantee it).  I think this would be more readable because then the code is operating more in terms of objects than indices into the different lists of subtokens and spans.

- iterate over subtokens, keeping track of current character index in the sentence (cumulative subtoken length)
- create a list variable to hold subtoken indices for the first/current span
- if current subtoken has any overlap with current span, add to list
- as soon as you are past the first span, add the list of found subtokens to the result and start a new list for the next span
- end when you hit the end of the last span

Reply:
I don't mean to get us off track with optimizing / refactoring we don't need for an experiment, just want to understand what is going on - if we're confident it is working as needed then there is no need to revise now! And I do think I understand what's going on well enough. But if we end up having to revise it for the experiment, refactoring might help.